### PR TITLE
Add `order` arg to stackplot

### LINKED ISSUE #266: Unexpected behaviour in `stack_plot()`
# Description of the issue

I've been trying to use the `stack_plot()` function for some follow-up [SR15](https://github.com/iiasa/ipcc_sr15_scenario_analysis) assessment, and noticed two major issues.

1) Timeseries data starting with zero are ignored completely
2) Multiple crossings of the zero baseline end up shown wrong

# Illustration

```
import pandas as pd
import pyam
df = pyam.IamDataFrame(pd.DataFrame([
    ['a', 1, 2, 3, 4],
    ['b', 0, 1, 2, 3],
    ['c', -1, 1, -1, 1],
    ['d', 1, 1, 1, -1]
    ],
    columns=['variable', 2010, 2020, 2030, 2040],
), model='model_a', scenario='scen_a', region='World', unit='some_unit')
df.stack_plot()
```

image

# More insights

In the source of `stack_plot()`, there is this docstring...

```
    # Sort lines so that negative timeseries are on the right, positive
    # timeseries are on the left and timeseries which go from positive to
    # negative are ordered such that the timeseries which goes negative first
    # is on the right (case of timeseries which go from negative to positive
    # is an edge case we haven't thought about as it's unlikely to apply to
    # us).
```

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR adds an `order` arg to the `stackplot` option. In preparation of that, it makes a few clean-up operations:
 - pass the IamDataFrame to the plotting library (rather than the data) to keep `core.py` shorter
 - deprecate `stack_plot()` in favor of `stackplot()` for consistency with Matplotlib
 - add a page with plotting docs (properly cross-referenced from the IamDataFrame docs)
 - add an `order` kwarg to `stackplot` with the behavior as explained below

The implementation of the ordering is slightly more complicated than necessary, anticipating that this will also be useful for ordering in other plotting functions. This new structure of the function names and docs can then be applied to the other plotting functions later.

The behaviour of the `order` kwarg:
 - if not specified, use the `run_control()['order']` dictionary of lists to order the stacked levels;
    any elements in the data which are not in the dictionary are added alphabetically after
 - if not specified and nothing relevant defined in the run-control: keep current (alphabetical) behavior
 - or use the order given by that kwarg if it is a list, dropping all elements in the data that are not in the list

This PR also fixes some weird behavior when several timeseries data cross the zero-axis multiple times.

closes #266

### CODE REVIEW

**Review by znicholls — APPROVED** [review]
lgtm

File: pyam/core.py [hunk 1]
```diff
@@ -72,8 +72,8 @@ class IamDataFrame(object):
 
     Parameters
     ----------
-    data : :class:`pandas.DataFrame`, :class:`ixmp.Scenario`,
-            or file-like object as str or :class:`pathlib.Path`
+    data : :class:`pandas.DataFrame`, :class:`ixmp.Scenario`,\
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/core.py:75]: I tried that, stickler doesn't like it because the line is too long... And the `\` follows the numpydoc convention, see https://github.com/numpy/numpydoc/issues/87

**Comment by gidden** [comment]
One suggested change - I think 'color' should be passed explicitly as the order if desired, as to not hide plot logic in a configuration file.

**Comment by danielhuppmann** [comment]
> One suggested change - I think 'color' should be passed explicitly as the order if desired, as to not hide plot logic in a configuration file.

Disagree for a number of reasons:

0. The whole point of the run control is to hide logic in a configuration file.
0. I imagine someone seeing this in a tutorial or code snippet and wondering what `order='color'` should mean - red before blue?
0. There is no reason to think that the current approach (alphabetical) is any less surprising to a user than the new default approach (run control plus alphabeting) - as a user, I might expect the order of the input dataframe (when casting to an IamDataFrame) to be the most logical order... At least, the new approach allows to explicitly set the order if the default does not make sense.

**Comment by gidden** [comment]
> * The whole point of the run control is to hide logic in a configuration file.
The point of the config file is not to hide logic, but rather to explicitly define configuration in one place. For example, we could add an `order` section there which can be looked up here so that a single order can be defined once and used in multiple places.

What is suggested here, if I understand, is rather to, by default, infer (implicitly) an ordering parameter through another, arguably unrelated parameter.

In general, we should be explicit instead of implicit.
> * I imagine someone seeing this in a tutorial or code snippet and wondering what `order='color'` should mean - red before blue?
Fair point. See above for a more explicit suggestion.

**Comment by danielhuppmann** [comment]
Well... I agree that adding an adding an `order` option would be more explicit, but on second thought, I don't think it's worth it:
 - the docstring would change from 
   ```python
    # levels are ordered based on the `run_control()['color']` dictionary
    ```
    to
    ```python
    # levels are ordered based on the `run_control()['order']` dictionary
    ```
 - it would just mean taking a dictionary of dictionaries and repeating it as dictionary of lists. In the [IPCC SR15 context](https://github.com/iiasa/ipcc_sr15_scenario_analysis/blob/390a1d9a1dde4ac667d946cb6167ce814099dd40/assessment/sr15_specs.yaml#L55), just 20 extra lines with no additional info - which could just as well be embedded in the order of the dictionaries...

**Comment by gidden** [comment]
Well not every user may be using their configuration files in the exact same way. [Here](https://github.com/gidden/ar6-wg1-ch6-emissions/blob/master/plotting.yaml) for example, if we were to pull as is, all plots would change. I think its worth being explicit in this case.

**Comment by danielhuppmann** [comment]
Implemented a separate `order` feature in the run control per suggestion by @gidden

**Comment by danielhuppmann** [comment]
While testing the ordered stackplot in my real-world use case, I noticed that the ordering was actually overridden by the quick-fix for dealing with data that crosses the zero line. So I fixed that one, too... Not necessarily the prettiest plots, but at least correct.

For illustration of the problem
```python
import pandas as pd
import pyam

TEST_STACKPLOT_DF = pd.DataFrame([
    ['var1', 1, -1, 0],
    ['var2', 1, 1, -2],
    ['var3', -1, 1, 1],
],
    columns=['variable', 2005, 2010, 2015],
)

df = pyam.IamDataFrame(TEST_STACKPLOT_DF, model='model_a', scenario='scen_a', region='World', unit='foo')
df.stackplot(total=True)
```

An illustration of what was wrong with the previous implementation:
 - legend not ordered (alphabetically by default)
 - var1 (green) in 2005 goes from +1 to -1
 - var3 (orange) goes to -2 in 2015, instead of +1
image

New implementation:
image
